ql/src/java/org/apache/hadoop/hive/ql/exec/repl/util/FileListStreamer.java

```diff
@@ -0,0 +1,150 @@
+  private volatile boolean completed = false;
+  private volatile boolean initialized = false;
+
+
+
+  public FileListStreamer(LinkedBlockingQueue<String> cache, Path backingFile, Configuration conf) throws IOException {
+    this.cache = cache;
+    this.backingFile = backingFile;
+    this.conf = conf;
+  }
+
+  private void lazyInit() throws IOException {
+    if (backingFileWriterInTest == null) {
+      FileSystem fs = FileSystem.get(backingFile.toUri(), conf);
+      backingFileWriter = new BufferedWriter(new OutputStreamWriter(fs.create(backingFile)));
+    } else {
+      backingFileWriter = backingFileWriterInTest;
+    }
+    initialized = true;
+    LOG.info("Initialized a file based store to save a list at: {}", backingFile);
+  }
+
+  public boolean isValid() {
+    return valid;
+  }
+
+  // Blocks for remaining entries to be flushed to file.
+  @Override
+  public void close() throws IOException {
+    signalTostop = true;
+    synchronized (COMPLETION_LOCK) {
+      while (motiveToWait()) {
+        try {
+          COMPLETION_LOCK.wait(TimeUnit.SECONDS.toMillis(TIMEOUT_IN_SECS));
+        } catch (InterruptedException e) {
+          // no-op
+        }
+      }
+    }
+    if (!isValid()) {
+      throw new IOException("File list is not in a valid state:" + backingFile);
+    }
+  }
+
+  private boolean motiveToWait() {
+    return !completed && valid;
+  }
+
+  @Override
+  public void run() {
+    try {
+      lazyInit();
+    } catch (IOException e) {
+      valid = false;
+      throw new RuntimeException("Unable to initialize the file list streamer", e);
+    }
+    boolean exThrown = false;
+    while (!exThrown && (!signalTostop || !cache.isEmpty())) {
+      try {
+        String nextEntry = cache.poll(TIMEOUT_IN_SECS, TimeUnit.SECONDS);
+        if (nextEntry != null) {
+          backingFileWriter.write(nextEntry);
+          backingFileWriter.newLine();
+          LOG.debug("Writing entry {} to file list backed by {}", nextEntry, backingFile);
+        }
+      } catch (Exception iEx) {
+        if (!(iEx instanceof InterruptedException)) {
+          // not draining any more. Inform the producer to avoid OOM.
+          valid = false;
+          LOG.error("Exception while saving the list to file " + backingFile, iEx);
+          exThrown = true;
+        }
+      }
+    }
+    try{
+      closeBackingFile();
+      completed = true;
+    } finally {
+      synchronized (COMPLETION_LOCK) {
+        COMPLETION_LOCK.notify();
+      }
+    }
+    LOG.info("Completed the file list streamer backed by: {}", backingFile);
+  }
+
+  private void closeBackingFile() {
+    try {
+      backingFileWriter.close();
+      LOG.debug("Closed the file list backing file: {}", backingFile);
+    } catch (IOException e) {
+      LOG.error("Exception while closing the file list backing file", e);
+      valid = false;
+    }
+  }
+
+  @VisibleForTesting
+  public static void setBackingFileWriterInTest(BufferedWriter bufferedWriter) {
```

Comment:
mock this instead of having test specific code

Reply:
How will thread work in mocked case?